lib/identity_key/identity_key.c

```diff
@@ -0,0 +1,189 @@
+#include <nrfx_nvmc.h>
+
+#include <zephyr/sys/printk.h>
+#include <zephyr/kernel.h>
+
+/* Safe zeroize function from Mbed TLS.
+ * This function can be replaced with memset_s with C11 support
+ */
+static void safe_zeroize(void *v, size_t n)
+{
+	volatile unsigned char *p = v;
+
+	while (n--) {
+		*p++ = 0;
+	}
+}
+
+static int generate_random_secp256r1_private_key(uint8_t *key_buff)
+{
+	psa_status_t status;
+	psa_key_handle_t key_handle;
+	size_t olen;
+
+	/* Initialize PSA Crypto */
+	status = psa_crypto_init();
+	if (status != PSA_SUCCESS) {
+		printk("Failure: Could not initialize PSA crypto!");
+		return -ERR_IDENTITY_KEY_INTERNAL_ERROR;
+	}
+
+	/* Configure the key attributes */
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+	/* Configure the key attributes for Curve type secp256r1
+	 * This key needs to be exported from the volatile storage
+	 */
+	psa_set_key_usage_flags(&key_attributes, PSA_KEY_USAGE_SIGN_HASH | PSA_KEY_USAGE_EXPORT);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&key_attributes, IDENTITY_KEY_SIZE_BYTES * 8);
+
+	status = psa_generate_key(&key_attributes, &key_handle);
+	if (status != PSA_SUCCESS) {
+		printk("psa_generate_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_GENERATE_FAILED;
+	}
+
+	status = psa_export_key(key_handle, key_buff, IDENTITY_KEY_SIZE_BYTES, &olen);
+	if (status != PSA_SUCCESS) {
+		printk("psa_export_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_EXPORT_FAILED;
+	}
+
+	if (olen != IDENTITY_KEY_SIZE_BYTES) {
+		printk("Exported key size failed! Key size: %d", olen);
+		return -ERR_IDENTITY_KEY_EXPORT_FAILED;
+	}
+
+	status = psa_destroy_key(key_handle);
+	if (status != PSA_SUCCESS) {
+		printk("psa_destroy_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_INTERNAL_ERROR;
+	}
+
+	return 0;
+}
+
+void identity_key_write_random(void)
+{
+	uint8_t key_buff[IDENTITY_KEY_SIZE_BYTES];
+	int err;
+
+	err = generate_random_secp256r1_private_key(key_buff);
+	if (err != 0) {
+		printk("The identity key generation failed with error code: %d", err);
+		/* Panic raised, as we can't generate the identity key! */
+		k_panic();
+	}
+
+	/* Write the key into KMU using the MKEK as encryption key */
+	identity_key_write_key(key_buff);
+
+	/* Clear the unencrypted key from RAM for security reasons */
+	safe_zeroize(key_buff, IDENTITY_KEY_SIZE_BYTES);
+}
+
+void identity_key_write_key(uint8_t key[IDENTITY_KEY_SIZE_BYTES])
+{
+	int err;
+
+	err = nrf_cc3xx_platform_identity_key_store(NRF_KMU_SLOT_KIDENT, key);
+
+	if (err != NRF_CC3XX_PLATFORM_SUCCESS) {
+		printk("The identity key writing to: %d", NRF_KMU_SLOT_KIDENT);
+		printk("failed with error code: %d", err);
+		/* Panic raised, as we can't write the generated identity key! */
+		k_panic();
+	}
+}
```

Comment:
Shouldn't we check that mkek is present first like the retrievel function does?

Reply:
Both of the functions check if the MKEK is written. I will add it to the write as well just to have better reporting.